bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+
+                // Compute and validate hash of the received key share
+                if let Err(api_error) =
+                    Self::validate_share(&from, &share, &share_digests)
+                {
+                    return api_error.into();
+                }
+
+                // Keep track of our new valid share
+                shares.insert(share);
+                from_all.insert(from);
+
+                // If we have enough shares, try to recompute the rack secret
+                // If we fail here, this is a major problem so we report it and
+                // clear our rack secret state, as we won't be able to succeed.
+                // This error should be effectively impossible to ever hit.
+                if shares.0.len() == *threshold {
+                    match RackSecret::combine_shares(&shares.0) {
+                        Ok(secret) => {
+                            *self = RackSecretState::Computed {
+                                secret: secret.clone(),
+                                expiry: rack_secret_expiry,
+                            };
+                            ApiOutput::RackSecret(secret).into()
+                        }
+                        Err(_) => {
+                            *self = RackSecretState::Empty;
+                            ApiError::FailedToReconstructRackSecret.into()
+                        }
+                    }
+                } else {
+                    // We don't have enough shares
+                    Output::none()
+                }
+            }
+            RackSecretState::Computed { .. } => {
+                // This is just a late share response. We've already informed
+                // the API caller and satisfied any pending learn requests.
+                Output::none()
+            }
+        }
+    }
+}
+
+/// Initialization state for tracking rack init on the coordinator node (the one
+/// local to RSS)
+#[derive(Debug)]
+pub enum RackInitState {
+    // This only exists on the RSS sled
+    Running {
+        start: Ticks,
+        total_members: usize,
+        acks: BTreeSet<Baseboard>,
+
+        // Any unacked messages will be re-sent when the destination peer
+        // connects.
+        unacked_envelopes: Vec<Envelope>,
+    },
+    Initialized,
+    Timeout,
+}
+
+impl RackInitState {
+    // Return true if we just completed rack init, false otherwise
+    pub fn on_ack(&mut self, from: Baseboard) -> bool {
+        if let RackInitState::Running {
+            total_members,
+            acks,
+            unacked_envelopes,
+            ..
+        } = self
+        {
+            unacked_envelopes.retain(|envelope| envelope.to != from);
+            acks.insert(from);
+            if acks.len() == *total_members {
+                *self = RackInitState::Initialized;
+                return true;
+            }
+        }
+        false
+    }
+    pub fn timer_expired(&self, now: Ticks, timeout: Ticks) -> bool {
```

Comment:
Add a docstring saying this returns true if your timer expired.

It also seems weird that we're passing in both `now` and `timeout`; since the former is changing and the latter is (presumably) a configuration value.

We could pass in `now - timeout`, which conveys the required information, but is hard to think about / name (`latest_start_time`?)

Reply:
Fixed in the `request_manager` in [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98).

We always use an expiry deadline now.